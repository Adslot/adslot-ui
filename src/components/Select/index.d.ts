import * as React from 'react';
import { Props, OptionTypeBase, components, createFilter } from 'react-select';
import CreatableSelect from 'react-select/creatable';
import AsyncSelect from 'react-select/async';
import AsyncCreatableSelect from 'react-select/async-creatable';

export interface SelectProps {
  dts?: string;
  isInModal?: boolean;
}

declare const Select: (<
  OptionType extends OptionTypeBase = { label: string; value: string },
  IsMulti extends boolean = false
>(
  props: SelectProps & Props<OptionType, IsMulti>
) => React.ReactElement<any, any> | null) & {
  components: typeof components;
  Creatable: typeof CreatableSelect;
  Async: typeof AsyncSelect;
  AsyncCreatable: typeof AsyncCreatableSelect;
  createFilter: typeof createFilter;
};

export default Select;
