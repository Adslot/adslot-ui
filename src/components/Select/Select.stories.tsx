import type { Meta, StoryObj } from '@storybook/react';

import Select from './index';

const meta = {
  title: 'Pending Review/Select',
  component: Select,
  tags: ['autodocs'],
} satisfies Meta<typeof Select>;

export default meta;

type Story = StoryObj<typeof meta>;

const countryOptions = [
  { value: 'au', label: 'Australia' },
  { value: 'ca', label: 'Canada' },
  { value: 'jp', label: 'Japan', disabled: true },
  { value: 'uk', label: 'United Kingdom' },
];

export const Default: Story = {
  args: {
    options: countryOptions,
    isInModal: true,
  },
};
