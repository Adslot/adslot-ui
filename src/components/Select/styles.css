@import url('../../styles/variable.css');

.select-component .caret-icon {
  width: 8px;
  height: 8px;
  background-repeat: no-repeat;
  background-size: contain;
  background-image: url('../../styles/icons/caret.svg');
}

.select-component .css-oeuhkd-control {
  border-color: $color-grey-300;
  border-radius: $border-radius-base;
}
