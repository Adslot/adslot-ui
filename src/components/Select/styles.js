import * as tokens from '../../../system/dist';
const { color } = tokens;
const borderColor = color.gray.base;

const defaultStyle = {
  option: (styles, { isFocused, isSelected, isDisabled, isActive }) => {
    let backgroundColor;
    switch (true) {
      case isSelected:
        backgroundColor = color.gray.lightest;
        break;
      case isFocused:
        backgroundColor = color.gray.lightest;
        break;
      case isDisabled:
        return styles;
      default:
        backgroundColor = null;
    }

    return {
      ...styles,
      color: 'inherit',
      ':active': { backgroundColor: color.gray.lighter },
      backgroundColor,
    };
  },
  indicatorSeparator: (styles) => ({
    ...styles,
    display: 'none',
  }),
  control: (styles, state) => ({
    ...styles,
    ...(state.isFocused
      ? {
          boxShadow: 0,
          ':hover': { borderColor },
          borderColor,
        }
      : {}),
    minHeight: 26,
    borderRadius: 0,
  }),
  clearIndicator: (styles) => ({
    ...styles,
    padding: '0 4px 0 8px',
  }),
  dropdownIndicator: (styles) => ({
    ...styles,
    padding: '0 8px 0 4px',
  }),
  valueContainer: (styles) => ({
    ...styles,
    padding: '0 8px',
  }),
  multiValue: (styles) => ({
    ...styles,
    color: '#fff',
    backgroundColor: color.gray.dark,
  }),
  multiValueLabel: (styles) => ({
    ...styles,
    color: '#fff',
  }),
  multiValueRemove: (styles) => ({
    ...styles,
    ':hover': {},
  }),
  menu: (styles) => ({
    ...styles,
    zIndex: 1060,
    borderRadius: 0,
    marginTop: 4,
  }),
  noOptionsMessage: (styles) => ({
    ...styles,
    textAlign: 'left',
  }),
};

export default defaultStyle;
