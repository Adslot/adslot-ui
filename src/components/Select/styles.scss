@import '~styles/mixin';
@import '~styles/variable';

.select-component {
  .caret-icon {
    @include svg-icon('~styles/icons/caret.svg', 8px, 8px);
  }

  .css-oeuhkd-control {
    border-color: var(--color-gray-lighter);
    border-radius: 2px;
  }
}
