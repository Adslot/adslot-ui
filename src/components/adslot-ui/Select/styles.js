const defaultStyle = {
  option: (styles, { isFocused, isSelected, isDisabled }) => {
    let backgroundColor;

    switch (true) {
      case isSelected:
        backgroundColor = '#f5faff';
        break;
      case isFocused:
        backgroundColor = '#f9f9f9';
        break;
      case isDisabled:
        return styles;
      default:
        backgroundColor = null;
    }

    return {
      ...styles,
      color: 'inherit',
      backgroundColor,
    };
  },
  indicatorSeparator: styles => ({
    ...styles,
    display: 'none',
  }),
  control: (styles, state) => ({
    ...styles,
    ...(state.isFocused
      ? {
          boxShadow: '0 0 0 1px #cccccc',
          ':hover': { borderColor: '#cccccc' },
          borderColor: '#cccccc',
        }
      : {}),
    minHeight: 26,
  }),
  clearIndicator: styles => ({
    ...styles,
    padding: '0 4px 0 8px',
  }),
  dropdownIndicator: styles => ({
    ...styles,
    padding: '0 8px 0 4px',
  }),
  valueContainer: styles => ({
    ...styles,
    padding: '0 8px',
  }),
  multiValue: styles => ({
    ...styles,
    color: '#fff',
    backgroundColor: '#838383',
  }),
  multiValueLabel: styles => ({
    ...styles,
    color: '#fff',
  }),
  multiValueRemove: styles => ({
    ...styles,
    ':hover': {},
  }),
  menu: styles => ({
    ...styles,
    zIndex: 1060,
  }),
};

export default defaultStyle;
