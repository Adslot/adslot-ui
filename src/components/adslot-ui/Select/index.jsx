import React from 'react';
import Select, { components } from 'react-select';
import caretIcon from '../../../styles/icons/caret.svg';
import defaultStyle from './styles';

const componentBaseClass = 'select-component';

const DropdownIndicator = props => (
  <components.DropdownIndicator {...props}>
    <img src={caretIcon} />
  </components.DropdownIndicator>
);

const ClearIndicator = props => <components.ClearIndicator {...props}>âœ•</components.ClearIndicator>;

// The propType for this component should be the same as 'React-Select` lib
const SelectComponent = props => (
  <Select
    components={{ DropdownIndicator, ClearIndicator }}
    className={componentBaseClass}
    classNamePrefix={componentBaseClass}
    styles={defaultStyle}
    {...props}
  />
);

export default SelectComponent;
