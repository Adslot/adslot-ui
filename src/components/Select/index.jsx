/* eslint-disable react/prop-types */
import _ from 'lodash';
import classnames from 'classnames';
import React from 'react';
import PropTypes from 'prop-types';
import ReactSelect, { components, createFilter } from 'react-select';
import CreatableSelect from 'react-select/creatable';
import AsyncSelect from 'react-select/async';
import AsyncCreatableSelect from 'react-select/async-creatable';
import defaultStyle from './styles';
import './styles.css';

const componentBaseClass = 'select-component';

const DropdownIndicator = (props) => (
  <components.DropdownIndicator {...props}>
    <div className="caret-icon" />
  </components.DropdownIndicator>
);

const ClearIndicator = (props) => <components.ClearIndicator {...props}>âœ•</components.ClearIndicator>;

const selectContainerBuilder = (extraProps) => (props) => {
  const containerProps = {
    ...props,
    innerProps: {
      ...props.innerProps,
      ...extraProps,
    },
  };

  return <components.SelectContainer {...containerProps} />;
};

/**
 * The propType for this component should be the same as 'React-Select` lib, plus any special props
 */
const selectComponentBuilder = (Component) => {
  const SelectComponent = (props) => {
    const customComponents = { DropdownIndicator, ClearIndicator };

    if (!_.isEmpty(props.dts)) {
      customComponents.SelectContainer = selectContainerBuilder({ 'data-test-selector': props.dts });
    }

    const selectProps = {
      ..._.omit(props, ['components', 'className', 'styles', 'dts', 'isInModal']),
      components: { ...customComponents, ...props.components },
      className: classnames(componentBaseClass, props.className),
      classNamePrefix: props.classNamePrefix || componentBaseClass,
      styles: { ...defaultStyle, ...props.styles },
    };

    if (props.isInModal) {
      selectProps.menuPortalTarget = document.body;
      selectProps.styles = {
        menuPortal: (base) => ({ ...base, zIndex: 9999 }),
        ...selectProps.styles,
      };
    }

    return <Component {...selectProps} />;
  };

  SelectComponent.propTypes = {
    dts: PropTypes.string,
    isInModal: PropTypes.bool,
  };
  SelectComponent.defaultProps = {
    isInModal: false,
  };

  return SelectComponent;
};

const Select = selectComponentBuilder(ReactSelect);

// re-export components so user can customize various components
Select.components = components;
Select.Creatable = selectComponentBuilder(CreatableSelect);
Select.Async = selectComponentBuilder(AsyncSelect);
Select.AsyncCreatable = selectComponentBuilder(AsyncCreatableSelect);
Select.createFilter = createFilter;

export default Select;
/* eslint-enable react/prop-types */
