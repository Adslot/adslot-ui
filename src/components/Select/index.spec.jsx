import React from 'react';
import { render, cleanup, fireEvent, queryAllByAttribute } from '@testing-library/react';
import Select from '.';

afterEach(cleanup);

const queryAllByDts = queryAllByAttribute.bind(null, 'data-test-selector');
const queryAllByClass = queryAllByAttribute.bind(null, 'class');

const defaultOptions = [
  { value: 'au', label: 'Australia' },
  { value: 'ca', label: 'Canada' },
  { value: 'jp', label: 'Japan' },
  { value: 'uk', label: 'United Kingdom' },
];

describe('<Select />', () => {
  it('should have base class', () => {
    const { queryAllByText } = render(<Select />);
    expect(queryAllByText('Select...')).toHaveLength(1);
  });

  it('should have custom option background', () => {
    const { getByText, queryAllByText } = render(
      <Select options={defaultOptions} isOptionDisabled={({ value }) => value === 'jp'} menuIsOpen />
    );

    expect(queryAllByText('Select...')).toHaveLength(1);
    expect(queryAllByText('Australia')).toHaveLength(1);
    expect(getByText('Australia')).toHaveClass('select-component__option');
    expect(queryAllByText('Canada')).toHaveLength(1);
    expect(getByText('Canada')).toHaveClass('select-component__option');
    expect(queryAllByText('Japan')).toHaveLength(1);
    expect(getByText('Japan')).toHaveClass('select-component__option');
    expect(queryAllByText('United Kingdom')).toHaveLength(1);
    expect(getByText('United Kingdom')).toHaveClass('select-component__option');

    fireEvent.click(getByText('Australia'));
    expect(queryAllByText('Australia')).toHaveLength(2);
    expect(queryAllByText('Australia')[0]).toHaveClass('select-component__single-value');
    expect(queryAllByText('Australia')[1]).toHaveClass(
      'select-component__option select-component__option--is-focused select-component__option--is-selected'
    );

    fireEvent.mouseOver(getByText('Canada'));
    expect(queryAllByText('Australia')[1]).not.toHaveClass('select-component__option--is-focused');
    expect(getByText('Canada')).toHaveClass('select-component__option--is-focused');

    expect(getByText('Japan')).toHaveClass('select-component__option--is-disabled');
  });

  it('should show cross icon for clear opration', () => {
    const { getByText, queryAllByText } = render(<Select options={defaultOptions} menuIsOpen isClearable />);
    fireEvent.click(getByText('Australia'));

    expect(queryAllByText('✕')).toHaveLength(1);
    expect(getByText('✕')).toHaveClass('select-component__indicator select-component__clear-indicator');
  });

  it('should handle multi selections', () => {
    const { container, getByText, queryAllByText } = render(
      <Select options={defaultOptions} defaultValue={[defaultOptions[0], defaultOptions[1]]} isMulti />
    );

    expect(queryAllByClass(container, 'css-11547y4 select-component__multi-value__label')).toHaveLength(2);
    expect(queryAllByText('Australia')).toHaveLength(1);
    expect(getByText('Australia')).toHaveClass('select-component__multi-value__label');
    expect(queryAllByText('Canada')).toHaveLength(1);
    expect(getByText('Canada')).toHaveClass('select-component__multi-value__label');
  });

  it('should have custom SelectContainer if dts is passed in', () => {
    const { container, rerender, getByText, queryAllByText } = render(
      <Select options={defaultOptions} dts="test-dts" />
    );

    expect(queryAllByDts(container, 'test-dts')).toHaveLength(1);
    expect(queryAllByClass(container, 'select-component__value-container css-9evde0')).toHaveLength(1);
    expect(
      queryAllByClass(
        container,
        'select-component__indicator select-component__dropdown-indicator css-171ess4-indicatorContainer'
      )
    ).toHaveLength(1);

    rerender(<Select options={defaultOptions} menuIsOpen isClearable dts="test-dts" />);
    fireEvent.click(getByText('Australia'));

    expect(queryAllByText('✕')).toHaveLength(1);
    expect(getByText('✕')).toHaveClass('select-component__indicator select-component__clear-indicator');
  });

  it('should render select options in body if props.isInModal is true', () => {
    const { container } = render(<Select options={defaultOptions} isInModal />);
    expect(container).toMatchSnapshot();
  });
});
