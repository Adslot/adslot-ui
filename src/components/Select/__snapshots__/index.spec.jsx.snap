// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should render select options in body if props.isInModal is true 1`] = `
<div>
  <div
    class="select-component css-2b097c-container"
  >
    <div
      class="select-component__control css-oeuhkd-control"
    >
      <div
        class="select-component__value-container css-9evde0"
      >
        <div
          class="select-component__placeholder css-1wa3eu0-placeholder"
        >
          Select...
        </div>
        <div
          class="css-b8ldur-Input"
        >
          <div
            class="select-component__input"
            style="display: inline-block;"
          >
            <input
              aria-autocomplete="list"
              autocapitalize="none"
              autocomplete="off"
              autocorrect="off"
              id="react-select-7-input"
              spellcheck="false"
              style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
              tabindex="0"
              type="text"
              value=""
            />
            <div
              style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
            />
          </div>
        </div>
      </div>
      <div
        class="select-component__indicators css-1hb7zxy-IndicatorsContainer"
      >
        <span
          class="select-component__indicator-separator css-43ykx9-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          class="select-component__indicator select-component__dropdown-indicator css-171ess4-indicatorContainer"
        >
          <div
            class="caret-icon"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
